src/mango_selector_text.erl

```diff
@@ -159,9 +159,17 @@ convert(Path, {[{<<"$size">>, Arg}]}) ->
 convert(_Path, {[{<<"$", _/binary>>=Op, _}]}) ->
     ?MANGO_ERROR({invalid_operator, Op});
 
-% We've hit a field name specifier. We need to break the name
-% into path parts and continue our conversion.
-convert(Path, {[{Field, Cond}]}) ->
+% We've hit a field name specifier. Check if the field name is accessing
+% arrays. Convert occurrences of element position references to .[]. Then we
+% need to break the name into path parts and continue our conversion.
+convert(Path, {[{Field0, Cond}]}) ->
+    Field = case re:run(Field0, <<"\\.\\d+">>) of
+        no_match ->
+            Field0;
+        _ ->
+            re:replace(Field0,<<"\\.\\d+">>,<<"\\.\\[\\]">>,[global,
+                {return,binary}])
+    end,
```

Comment:
I'm disliking that there are three uncompiled regexps when we have to make this change. And the old single regexp version looks broken anyway.

https://github.com/cloudant/mango/blob/46268-fix-match-arrays/src/mango_doc.erl#L540-L564

I'd move those mango_doc functions to mango_util, then parse the field once here adding a conversion function something like:

    convert(Path, {[{Field0, Cond}]}) ->
        {ok, PathParts0} = mango_util:parse_field(Field0),
        PathParts1 = replace_array_indexes(Path0),
        NewPath = lists:reverse(PathParts1) ++ Path,
        convert(NewPath, Cond);


    replace_array_indexes([]) ->
        [];
    replace_array_indexes([Part | Rest]) ->
        NewParts = try
            list_to_integer(binary_to_list(Part)),
            <<"[]">>;
        catch _:_ ->
            Part
        end,
        [NewPart | replace_array_indexes(Rest)].